File: src/css/properties/prefix_handler.zig
const std = @import("std");
const bun = @import("root").bun;
const Allocator = std.mem.Allocator;

pub const css = @import("../css_parser.zig");

const CustomPropertyName = css.css_properties.CustomPropertyName;

const Printer = css.Printer;
const PrintErr = css.PrintErr;
const VendorPrefix = css.VendorPrefix;
const Error = css.Error;

const PropertyId = css.PropertyId;
const PropertyIdTag = css.PropertyIdTag;
const Property = css.Property;
const UnparsedProperty = css.css_properties.custom.UnparsedProperty;

/// *NOTE* The struct field names must match their corresponding names in `Property`!
pub const FallbackHandler = struct {
    color: ?usize = null,
    @"text-shadow": ?usize = null,
    // TODO: add these back plz
    // filter: ?usize = null,
    // @"backdrop-filter": ?usize = null,
    // fill: ?usize = null,
    // stroke: ?usize = null,
    // @"caret-color": ?usize = null,
    // caret: ?usize = null,

    const field_count = @typeInfo(FallbackHandler).Struct.fields.len;

    pub fn handleProperty(
        this: *FallbackHandler,
        property: *const Property,
        dest: *css.DeclarationList,
        context: *css.PropertyHandlerContext,
    ) bool {
        inline for (std.meta.fields(FallbackHandler)) |field| {
            if (@intFromEnum(@field(PropertyIdTag, field.name)) == @intFromEnum(@as(PropertyIdTag, property.*))) {
                const has_vendor_prefix = comptime PropertyIdTag.hasVendorPrefix(@field(PropertyIdTag, field.name));
                var val = if (comptime has_vendor_prefix)
                    @field(property, field.name)[0].deepClone(context.allocator)
                else
                    @field(property, field.name).deepClone(context.allocator);

                if (@field(this, field.name) == null) {
                    const fallbacks = val.getFallbacks(context.allocator, context.targets);
                    const has_fallbacks = !fallbacks.isEmpty();

                    for (fallbacks.slice()) |fallback| {
                        dest.append(
                            context.allocator,
                            @unionInit(
                                Property,
                                field.name,
                                if (comptime has_vendor_prefix)
                                    .{ fallback, @field(property, field.name)[1] }
                                else
                                    fallback,
                            ),
                        ) catch bun.outOfMemory();
                    }
                    if (comptime has_vendor_prefix) {
                        if (has_fallbacks and @field(property, field.name[1]).contains(VendorPrefix{ .none = true })) {
                            @field(property, field.name[1]) = css.VendorPrefix{ .none = true };
                        }
                    }
                }

                if (@field(this, field.name) == null or
                    context.targets.browsers != null and !val.isCompatible(context.targets.browsers.?))
                {
                    @field(this, field.name) = dest.items.len;
                    dest.append(
                        context.allocator,
                        @unionInit(
                            Property,
                            field.name,
                            if (comptime has_vendor_prefix)
                                .{ val, @field(property, field.name)[1] }
                            else
                                val,
                        ),
                    ) catch bun.outOfMemory();
                } else if (@field(this, field.name) != null) {
                    const index = @field(this, field.name).?;
                    dest.items[index] = @unionInit(
                        Property,
                        field.name,
                        if (comptime has_vendor_prefix)
                            .{ val, @field(property, field.name)[1] }
                        else
                            val,
                    );
                } else {
                    val.deinit(context.allocator);
                }

                return true;
            }
        }

        if (@as(PropertyIdTag, property.*) == .unparsed) {
            const val: *const UnparsedProperty = &property.unparsed;
            var unparsed, const index = unparsed_and_index: {
                inline for (std.meta.fields(FallbackHandler)) |field| {
                    if (@intFromEnum(@field(PropertyIdTag, field.name)) == @intFromEnum(val.property_id)) {
                        const has_vendor_prefix = comptime PropertyIdTag.hasVendorPrefix(@field(PropertyIdTag, field.name));
                        const newval = newval: {
                            if (comptime has_vendor_prefix) {
                                if (@field(val.property_id, field.name)[1].contains(VendorPrefix{ .none = true }))
                                    break :newval val.getPrefixed(context.targets, @field(css.prefixes.Feature, field.name));
                            }
                            break :newval val.deepClone(context.allocator);
                        };
                        break :unparsed_and_index .{ newval, &@field(this, field.name) };
                    }
                }
                return false;
            };

            context.addUnparsedFallbacks(&unparsed);
            if (index.*) |i| {
                dest.items[i] = Property{ .unparsed = unparsed };
            } else {
                index.* = dest.items.len;
                dest.append(context.allocator, Property{ .unparsed = unparsed }) catch bun.outOfMemory();
            }

            return true;
        }

        return false;
    }

    pub fn finalize(this: *FallbackHandler, _: *css.DeclarationList, _: *css.PropertyHandlerContext) void {
        inline for (std.meta.fields(FallbackHandler)) |field| {
            @field(this, field.name) = null;
        }
    }
};

File: src/css/properties/prefix_handler.rs
use crate as css;
use crate::css_properties::custom::UnparsedProperty;
use crate::css_properties::{Property, PropertyIdTag};
use crate::generics::{DeepClone as _, IsCompatible as _};
use bun_alloc::ArenaVecExt as _;

/// *NOTE* The struct field names must match their corresponding variants in `Property`!
#[derive(Default)]
pub struct FallbackHandler {
    pub color: Option<usize>,
    pub text_shadow: Option<usize>,
    // The remaining fallback fields are not implemented yet.
    // filter: Option<usize>,
    // backdrop_filter: Option<usize>,
    // fill: Option<usize>,
    // stroke: Option<usize>,
    // caret_color: Option<usize>,
    // caret: Option<usize>,
}

impl FallbackHandler {
    pub(crate) fn handle_property(
        &mut self,
        property: &Property,
        dest: &mut css::DeclarationList,
        context: &mut css::PropertyHandlerContext,
    ) -> bool {
        // Each (field, Property variant, has_vendor_prefix) pair is expanded via macro.

        let arena = dest.bump();

        // Each (field, Property variant) pair is expanded via a macro that
        // takes per-type closures for `DeepClone`/`IsCompatible`/`get_fallbacks`,
        // letting each payload type use its own inherent methods.
        macro_rules! handle_unprefixed {
            (
                $self_field:ident,
                $Variant:ident,
                deep_clone = $dc:expr,
                fallbacks  = $fb:expr,
                is_compat  = $ic:expr
            ) => {
                if let Property::$Variant(payload) = property {
                    let mut val = ($dc)(payload, arena);

                    if $self_field.is_none() {
                        // `has_fallbacks` is only consulted in the vendor-prefixed branch.
                        ($fb)(&mut val, arena, &context.targets, dest);
                    }

                    if $self_field.is_none()
                        || (context.targets.browsers.is_some()
                            && !($ic)(&val, &context.targets.browsers.unwrap()))
                    {
                        *$self_field = Some(dest.len());
                        dest.push(Property::$Variant(val));
                    } else if let Some(index) = *$self_field {
                        dest[index] = Property::$Variant(val);
                    } else {
                        // val dropped — Drop handles cleanup
                        drop(val);
                    }

                    return true;
                }
            };
        }

        // Reshaped for borrowck — pre-borrow each self.<field> as &mut so the
        // macro body can both read and assign it without re-borrowing `self`.
        let this = &mut *self;
        let color = &mut this.color;
        let text_shadow = &mut this.text_shadow;

        // PropertyIdTag::Color has no vendor prefix.
        handle_unprefixed!(
            color,
            Color,
            deep_clone = |c: &css::css_values::color::CssColor, a| c.deep_clone(a),
            fallbacks = |v: &mut css::css_values::color::CssColor,
                         a: &bun_alloc::Arena,
                         t,
                         d: &mut css::DeclarationList| {
                let fbs = v.get_fallbacks(a, t);
                for fb in fbs.to_owned_slice().into_vec() {
                    d.push(Property::Color(fb));
                }
            },
            is_compat = |v: &css::css_values::color::CssColor, b| v.is_compatible(b)
        );
        // PropertyIdTag::TextShadow has no vendor prefix.
        handle_unprefixed!(
            text_shadow,
            TextShadow,
            deep_clone =
                |l: &css::SmallList<css::css_properties::text::TextShadow, 1>, a| l.deep_clone(a),
            fallbacks = |v: &mut css::SmallList<css::css_properties::text::TextShadow, 1>,
                         a: &bun_alloc::Arena,
                         t,
                         d: &mut css::DeclarationList| {
                for fb in css::small_list::get_fallbacks_text_shadow(v, a, t)
                    .to_owned_slice()
                    .into_vec()
                {
                    d.push(Property::TextShadow(fb));
                }
            },
            is_compat = |v: &css::SmallList<css::css_properties::text::TextShadow, 1>, b| v
                .is_compatible(b)
        );

        if let Property::Unparsed(val) = property {
            let val: &UnparsedProperty = val;
            let (mut unparsed, index): (UnparsedProperty, &mut Option<usize>) = 'unparsed_and_index: {
                macro_rules! match_unparsed_unprefixed {
                    ($self_field:ident, $Variant:ident) => {
                        if val.property_id.tag() == PropertyIdTag::$Variant {
                            let newval = val.deep_clone(arena);
                            break 'unparsed_and_index (newval, $self_field);
                        }
                    };
                }

                match_unparsed_unprefixed!(color, Color);
                match_unparsed_unprefixed!(text_shadow, TextShadow);
                // (no prefixed properties active yet — `match_unparsed_prefixed!` kept for
                // when filter/backdrop_filter/etc. are re-enabled in this handler.)

                return false;
            };

            context.add_unparsed_fallbacks(arena, &mut unparsed);
            if let Some(i) = *index {
                dest[i] = Property::Unparsed(unparsed);
            } else {
                *index = Some(dest.len());
                dest.push(Property::Unparsed(unparsed));
            }

            return true;
        }

        false
    }

    pub(crate) fn finalize(
        &mut self,
        _dest: &mut css::DeclarationList,
        _context: &mut css::PropertyHandlerContext,
    ) {
        self.color = None;
        self.text_shadow = None;
    }
}
